README

This is a long-dead fork. Active development is at https://github.com/ingwarsw/tinyfugue,
which now supports MSDP.  Thanks to Michał Lipka for reworking my changes into
the modern codebase and I hope someone finds it useful.

  TinyFugue is Copyright (C) 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2002, 2003, 2004, 2005, 2006-2007 Ken Keys.

  PCRE regexp package is Copyright 1997-1999 University of Cambridge.
  See src/pcre-2.08/LICENCE for details.

=======

  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

=======


                       INSTALLING TINYFUGUE

TinyFugue can be installed and run on UNIX-like systems, Mac OS X, OS/2,
and Win32.  For instructions, see the README file in the appropriate
directory.  For information on unofficial versions of TF for other
operating systems, see the TF web page at http://tinyfugue.sourceforge.net/.

Note that the .tar.gz and .tar.Z distributions contain LF line delimiters,
intended for UNIX, and the .zip distribution contains CR LF delimiters,
intended for OS/2.  It is possible to convert line delimiters, but it's
easier to just get a distribution that already has the correct delimiters.


"Make" options
--------------

Options are not available on all systems.  To use an option, give it as
an argument to the installation program (e.g., "make clean").

 all		Compile, but do not install (the "install" option should be
			run later to install the files).
 install	Compile (if you haven't already) and install.  The default
			installation locations depend on the operating system.
 clean		Remove object files and other junk from source directory.
 uninstall	Remove tf executable, help files, and library from their
			installed locations.


Public Installation
-------------------

Some features of TF can be disabled for secure public installation, by
using one of these /restrict commands in %{TFLIBDIR}/local.tf:

    /restrict SHELL  Prevents all access to shell or external commands.
                     Disables TF builtins "/sh" and "/quote !", and
                     uncompression during /load and /help.

    /restrict FILE   Prevents reading and writing of files.  Disables
                     TF builtins "/load", "/save", "/saveworld", "/log",
                     and "/quote '", and sockmload feature.
                     Implies /restrict shell.

    /restrict WORLD  Prevents the user from defining new worlds and
                     connecting to undefined worlds.  TF builtins
                     /addworld and the "/world <host> <port>" semantics
                     are disabled.  Implies /restrict file.


Terminal Handling
-----------------

If the default terminal handling option does not work, tf can be
configured to use vt100 codes or nothing at all.  See the README
file in the subdirectory corresponding to your operating system.


Compression
-----------

If you are short on disk space, you can compress the helpfile and
library files (except stdlib.tf).  Make sure the COMPRESS_SUFFIX
and COMPRESS_READ macros are set correctly; set them in the
%{TFLIBDIR}/local.tf file if needed.  Note that compresion can not
be used if /restrict is used.


Firewalls
---------

TF can be made to connect through a generic proxy server by setting
the %proxy_host variable at runtime.  See "/help proxy".

Also see the README file in the subdirectory corresponding to your
operating system to see if tf has transparent firewall support on your
system.


Last Resort
-----------

If you have an installation problem or other system-specific problem
that is not described in this README or the README in the subdirectory
for your system, contact the person who supports TF on your system.  If
you have a problem that is not system-related, ask for support on
http://tinyfugue.sourceforge.net/ or contact the author by email at
kenkeys@users.sourceforge.net.  Please provide the following information:

    The version of TF (type "/version" in tf).
    The operating system version (on unix, type "uname -a" in the shell).
    If tf won't compile, send the output of configure and make (in plaintext
        form, please).
    If you have a bug or core, give ALL error messages from tf.
    If you have a bug or core, describe what you did or what happened
        before the problem, and if the problem is repeatable.
    If you have a core file, do NOT send it.


Porting
-------

I have no personal interest in Mac OS 9, DOS, Windows 95, Windows NT, or
Amiga, so I don't plan on porting to them.  However, I am willing to
work with someone who wants to try and is willing to follow my
guidelines (this is how the OS/2 version was created, and why it is an
official version which continues to grow).  Basically, this means
keeping the platform-specific code seperate and to a minimum, and not
adding new features.



CHANGES
Only the most recent version will have detailed changes described.
Some fixes are followed by the version in which the problem first appeared.

5.0
---
If you have not yet read all changes since 4.0 stable 1, please do now.
  (read the CHANGES file directly, or type "/changes 5.0" in tf.)
If you have custom code to do any of the following in older versions,
  you'll probably want to get rid of it or redesign it, since it is now
  superceded by standard features: key bindings; text redraw (after world
  switch, or resuming from sh or suspend); scrollback; status line
  customizations; ...

5.0 beta 8+MSDP (2012-08-03)
----------
Negotiates MSDP with remote server.
Added /msdp command to send MSDP frames to server
MSDP responses cause triggers and update variables

5.0 beta 8 (2007-01-14)
----------
Added %max_hook to prevent infinite hook loops.
Added gethostname().
Added missing documentation for new "E" and "W" attributes. (50b7)
Added "inline" flag to decode_attr().
Added -a<attr> and -p options to /prompt.
Output of "/load testcolor.tf" now includes labels with r,g,b values.
May create a debugging log if tf crashes.
When invoked as a login shell, suspend is ignored, instead of hanging parent
    (workaround for sshd, which does not correctly set TSTP signal handler).
Fixed: after a resize with visual=off, the screen was not redrawn. (50b7)
Fixed: after a resize with visual=off, an immediate backspace or other input
    deletion would crash tf: "Internal error: output.c, line 2501". (50b7)
Fixed: server display attributes could be lost if they had wrapped from a
    previous line and were followed by more attributes.
Fixed: /filter changed the idea of the virtual bottom line, so a subsequent
    /unfilter redrew from the new incorrect bottom, and a subsequent
    /filter used the new incorrect bottom as a starting point for applying
    a new filter (resulting in a spurious "no lines match criteria" if the
    only matching lines are between the correct and incorrect bottoms).
Fixed potential crashes in calling functions with invalid arguments:
    strrep(), status_fields().
Fixed: on servers with terminated (GOAHEAD or EOR) prompts, tf could lose
    the first part of a server line in the rare condition of tf sending a
    line between receiving the first and second parts of the server line.
Fixed: non-time decimal values greater than 60 were sometimes formatted as
    time values. (50b7)
Fixed fatal bug in moresize() on a world that has never been /connect'ed.
Fixed infinite loop when replace() is called with empty first argument.
Fixed: with visual=off, /dokey flush (^[j) did not clear the more prompt.
Fixed: /recordline ignored -a option if -p was not also given.
Fixed fatal bug on some systems in /listsockets when there are connectionless
    sockets open.
Fixed: various bad things could happen if a trigger attempted to apply a
    partial hilite to a line longer than 32767 characters.
Fixed bugs on 64-bit platforms: /fg -c (Debian bug #327448), some variable
    assignments
Fixed: some cases of hostname resolution failure could cause a crash
More robust TELNET protocol handling reduces problems on noncompliant servers:
    Ignores option subnegotiation if option has not been agreed on.
    If IAC SB (0xff 0xfa) is received before any other telnet command, the
	server is treated as a non-telnet server.
Upgraded configure.in to autoconf 2.59 (should affect developers only)


5.0 beta 7 (2005-04-17)
----------
Colors:
    Added 16-color names: "gray", "brightred"..."brightwhite".
    Added 256-color names: "rgb000"..."rgb555" and "gray0"..."gray23".
    Added predefined codes for xterm 256-color extension.
    Functions decode_ansi() and encode_ansi() work with xterm 256-color
	extension codes.
    Added "/feature 256colors".
    Added "/load testcolor.tf" to test color functionality and appearance.
Added "E" and "W" attributes, with corresponding defintions in %error_attr
    and %warning_attr.
Added encode_ansi() to generate a string with attribute control codes.
Added test() for backward compatiblity with version 4.0.
Added strcmpattr(), like strcmp() but also compares display attributes.
In commands that match macros (/list, /purge, /save), a color attribute will
    match only the same color attribute, not just any color attribute.
/list, /listvar, /listworlds, and /listsockets can sort their results.
Added -n option to /listsockets to print host and port in numeric form.
%Pn, %L, and %R subs are now case sensitive, to reduce collisions with regular
    variables.
/prompt preserves embedded display attributes in its argument.
CONNECT hook includes SSL cipher name.
Added -f (foreground) and -b (background) options to /world; documented -x
    (SSL) option.
Uses system's copy of libpcre if possible.
Redefinition of macros or keybindings does not invoke the REDEF hook if the
    new and old defintions are identical (so file reloading is less noisy).
    If you used to gag REDEF to reduce noise, you may want to ungag it now
    that it is less noisy and more informative.
Status bar changes:
    Status bar can be multiple rows, determined by %status_height.
    /status_{add,edit,rm} commands take an -r option to specify which row.
    Added status_fields(<row>) to get fields of <row>.
    Added /status_defaults to restore default status bar fields; use of
	%status_field_defaults is deprecated.
    Added "/status_save <name>" and "/status_restore <name>" to save and
	restore a named set of status fields.
    The variable width field may be right-justified by giving it width "-0".
    Fixed: /status_edit did not work on a variable-width field.
    Fixed: "/status_add -x name:width:attr" failed to detect existing field
	with name if its width or attr were different.
    Fixed: status_attr_{int,var}_* variables sometimes had no effect. (50b5)
    Fixed: any attempt to set %status_pad in .tfrc failed if %visual had not

src/main.c
/*************************************************************************
 *  TinyFugue - programmable mud client
 *  Copyright (C) 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2002, 2003, 2004, 2005, 2006-2007 Ken Keys
 *
 *  TinyFugue (aka "tf") is protected under the terms of the GNU
 *  General Public License.  See the file "COPYING" for details.
 ************************************************************************/
static const char RCSid[] = "$Id: main.c,v 35004.115 2007/01/13 23:12:39 kkeys Exp $";


/***********************************************
 * Fugue main routine                          *
 *                                             *
 * Initializes many internal global variables, *
 * determines initial world (if any), reads    *
 * configuration file, and calls main loop in  *
 * socket.c                                    *
 ***********************************************/

#include "tfconfig.h"
#include "port.h"
#include "tf.h"
#include "util.h"
#include "pattern.h"	/* for tfio.h */
#include "search.h"
#include "tfio.h"
#include "history.h"
#include "world.h"
#include "socket.h"
#include "macro.h"
#include "output.h"
#include "attr.h"
#include "signals.h"
#include "command.h"
#include "keyboard.h"
#include "variable.h"
#include "tty.h"	/* no_tty */
#include "expand.h"
#include "expr.h"
#include "process.h"

const char sysname[] = UNAME;

/* For customized versions, please add a unique identifer (e.g., your initials)
 * to the version number, and put a brief description of the modifications
 * in the mods[] string.
 */
const char version[] =
#if DEVELOPMENT
    "DEVELOPMENT VERSION: "
#endif
    "TinyFugue version 5.0 beta 8+MSDP";

const char mods[] = "MSDP support written by Harik <git@harik.welp.gs>";

const char copyright[] =
    "Copyright (C) 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2002, 2003, 2004, 2005, 2006-2007 Ken Keys (kenkeys@users.sourceforge.net)";

const char contrib[] =
#ifdef PLATFORM_OS2
    "OS/2 support written by Andreas Sahlbach (asa@stardiv.de)\r"
#else
	"";
#endif


int restriction = 0;
int debug = 0;

static void read_configuration(const char *fname);
int main(int argc, char **argv);

int main(int argc, char *argv[])
{
    char *opt, *argv0 = argv[0];
    char *configfile = NULL, *command = NULL, *libdir = NULL;
    int worldflag = TRUE;
    int autologin = -1, quietlogin = -1, autovisual = TRUE;
    String *scratch;

    puts("");
    puts(version);
    puts(copyright);

    while (--argc > 0 && (*++argv)[0] == '-') {
        if (!(*argv)[1]) { argc--; argv++; break; }
        for (opt = *argv + 1; *opt; )
            switch (*opt++) {
            case 'd':
                debug = TRUE;
                break;
            case 'l':
                autologin = FALSE;
                break;
            case 'q':
                quietlogin = TRUE;
                break;
            case 'n':
                worldflag = FALSE;
                break;
            case 'v':
                autovisual = FALSE;
                break;
            case 'f':
                configfile = opt;
                goto nextarg;
            case 'c':
                command = opt;
                goto nextarg;
            case 'L':
                libdir = opt;
                goto nextarg;
            default:
                fprintf(stderr, "\n%s: illegal option -- %c\n", argv0, *--opt);
                goto error;
            }
        nextarg: /* empty statement */;
    }
    if (argc > 2) {
    error:
        fputs("\n", stderr);
        fprintf(stderr, "Usage: %s [-L<dir>] [-f[<file>]] [-c<cmd>] [-vnlq] [<world>]\n", argv0);
        fprintf(stderr, "       %s [-L<dir>] [-f[<file>]] [-c<cmd>] [-vlq] <host> <port>\n", argv0);
        fputs("Options:\n", stderr);
        fputs("  -L<dir>   use <dir> as library directory (%TFLIBDIR)\n", stderr);
        fputs("  -f        don't load personal config file (.tfrc)\n", stderr);
        fputs("  -f<file>  load <file> instead of config file\n", stderr);
        fputs("  -c<cmd>   execute <cmd> after loading config file\n", stderr);
        fputs("  -n        no automatic first connection\n", stderr);
        fputs("  -l        no automatic login/password\n", stderr);
        fputs("  -q        quiet login\n", stderr);
        fputs("  -v        no automatic visual mode\n", stderr);
        fputs("Arguments:\n", stderr);
        fputs("  <host>    hostname or IP address\n", stderr);
        fputs("  <port>    port number or name\n", stderr);
        fputs("  <world>   connect to <world> defined by addworld()\n", stderr);
        fputs("\n", stderr);
        exit(1);
    }

#if SOCKS
    SOCKSinit(argv0);  /* writes message to stdout */
#endif
    init_exename(argv0);

    SRAND(getpid() + time(NULL));	/* seed random generator */
    init_malloc();			/* malloc.c   */
    init_tfio();			/* tfio.c     */
    init_util1();			/* util.c     */
    init_expand();			/* expand.c   */
    init_variables();			/* variable.c */
    init_signals();			/* signals.c  */
    init_sock();			/* socket.c   */
    init_macros();			/* macro.c    */
    init_histories();			/* history.c  */
    init_output();			/* output.c   */
    init_attrs();			/* attr.c     */
    init_keyboard();			/* keyboard.c */

    oputs(version);
    oputs(copyright);
    oputs("Type `/help copyright' for more information.");
    if (*contrib) oputs(contrib);
    if (*mods) oputs(mods);
    oprintf("Using PCRE version %s", pcre_version());
    oputs("Type `/help', `/help topics', or `/help intro' for help.");
    oputs("Type `/quit' to quit tf.");
    oputs("");

    init_util2();			/* util.c     */

    if (libdir) {
        set_str_var_by_name("TFLIBDIR", Stringnew(libdir, -1, 0));
    }
    if (!ffindglobalvar("TFLIBRARY")) {
        scratch = Stringnew(NULL, 0, 0);
        Sprintf(scratch, "%s/stdlib.tf", TFLIBDIR);
        set_str_var_by_name("TFLIBRARY", scratch);
    }
    if (!ffindglobalvar("TFHELP")) {
        scratch = Stringnew(NULL, 0, 0);
        Sprintf(scratch, "%s/tf-help", TFLIBDIR);
        set_str_var_by_name("TFHELP", scratch);
    }

    read_configuration(configfile);

    if (command) {
	String *scmd;
	(scmd = Stringnew(command, -1, 0))->links++;
        macro_run(CS(scmd), 0, NULL, 0, sub, "\bSTART");
	Stringfree(scmd);
    }

    /* If %visual was not explicitly set, set it now. */
    if (getintvar(VAR_visual) < 0 && !no_tty)
        set_var_by_id(VAR_visual, autovisual);

    /* If %interactive was not explicitly set, set it now. */
    if (getintvar(VAR_interactive) < 0)
        set_var_by_id(VAR_interactive, !no_tty);

    if (argc > 0 || worldflag) {
	int flags = 0;
        if (autologin < 0) autologin = login;
        if (quietlogin < 0) quietlogin = quietflag;
	if (autologin) flags |= CONN_AUTOLOGIN;
	if (quietlogin) flags |= CONN_QUIETLOGIN;
        if (argc == 0)
            openworld(NULL, NULL, flags);
        else if (argc == 1)
            openworld(argv[0], NULL, flags);
        else /* if (argc == 2) */
            openworld(argv[0], argv[1], flags);
    } else {
        world_hook("---- No world ----", NULL);
    }

    main_loop();

    kill_procs();
#if USE_DMALLOC
    free_screen_lines(default_screen);
#endif
    minimal_fix_screen();
    reset_tty();

#if USE_DMALLOC
    free_macros();
    free_worlds();
    free_histories();
    free_output();
    free_vars();
    free_keyboard();
    free_search();
    free_expand();
    free_expr();
    free_help();
    free_util();
    free_patterns();
    free_dstring();
    free_reserve();
    fflush(stdout);
    fflush(stderr);
    debug_mstats("tf");
#endif

    return 0;
}

static void read_configuration(const char *fname)
{
#if 1 /* XXX */
    if (do_file_load(getvar("TFLIBRARY"), FALSE) < 0)
        die("Can't read required library.", 0);
#endif

    if (fname) {
        if (*fname) do_file_load(fname, FALSE);
        return;
    }

    (void)(   /* ignore value of expression */
	/* Try the next file if a file can't be read, but not if there's
	 * an error _within_ a file. */
        do_file_load("~/.tfrc", TRUE) >= 0 ||
        do_file_load("~/tfrc",  TRUE) >= 0 ||
        do_file_load("./.tfrc", TRUE) >= 0 ||
        do_file_load("./tfrc",  TRUE)
    );

    /* support for old fashioned .tinytalk files */
    do_file_load((fname = getvar("TINYTALK")) ? fname : "~/.tinytalk", TRUE);
}



src/varlist.h
/*************************************************************************
 *  TinyFugue - programmable mud client
 *  Copyright (C) 1996, 1997, 1998, 1999, 2002, 2003, 2004, 2005, 2006-2007 Ken Keys
 *
 *  TinyFugue (aka "tf") is protected under the terms of the GNU
 *  General Public License.  See the file "COPYING" for details.
 ************************************************************************/
/* $Id: varlist.h,v 35000.78 2007/01/13 23:12:39 kkeys Exp $ */

/* This keeps the constants and the array in the same place, so they can't
 * get out of sync.
 */

#define varstr(id, name, val, func) \
    varcode(id, name, val, TYPE_STR, 0, NULL, 0, 0, func)
#define varstrx(id, name, val, func) \
    varcode(id, name, val, TYPE_STR, VARAUTOX, NULL, 0, 0, func)
#define varint(id, name, ival, func) \
    varcode(id, name, NULL, TYPE_INT, 0, NULL, ival, 0, func)
#define vartime(id, name, ival, uval, func) \
    varcode(id, name, NULL, TYPE_DTIME, 0, NULL, ival, uval, func)
#define varpos(id, name, ival, func) \
    varcode(id, name, NULL, TYPE_POS, 0, NULL, ival, 0, func)
#define varenum(id, name, ival, func, enums) \
    varcode(id, name, NULL, TYPE_ENUM, 0, enums, ival, 0, func)
#define varflag(id, name, ival, func) \
    varcode(id, name, NULL, TYPE_ENUM, 0, enum_flag, ival, 0, func)

/* These MUST be sorted by name! */
/*      ID,		NAME,		VALUE,		FUNC,	ENUMS	*/

#if HAVE_SETLOCALE
varstrx(VAR_LANG,	"LANG",		NULL,		ch_locale)
varstrx(VAR_LC_ALL,	"LC_ALL",	NULL,		ch_locale)
varstrx(VAR_LC_CTYPE,	"LC_CTYPE",	NULL,		ch_locale)
varstrx(VAR_LC_TIME,	"LC_TIME",	NULL,		ch_locale)
#endif /* HAVE_SETLOCALE */
varstr (VAR_MAIL,	"MAIL",		NULL,		ch_mailfile)
varstr (VAR_TERM,	"TERM",		NULL,		change_term)
varstr (VAR_TFLIBDIR,	"TFLIBDIR",	DEFAULT_TFLIBD,	NULL)
varstr (VAR_TFMAILPATH,	"TFMAILPATH",	NULL,		ch_mailfile)
varstr (VAR_TFPATH,	"TFPATH",	NULL,		NULL)
varstrx(VAR_TZ,		"TZ",		NULL,		ch_timezone)
varstr (VAR_alert_attr,	"alert_attr",	"Br",		ch_attr)
vartime(VAR_alert_time,	"alert_time",	5,0,		NULL)
#if 0
varflag(VAR_auto_fg,	"auto_fg",	FALSE,		NULL)
#endif
varflag(VAR_background,	"background",	TRUE,		tog_bg)
varflag(VAR_backslash,	"backslash",	TRUE,		NULL)
varenum(VAR_bamf,	"bamf",		FALSE,		NULL,	enum_bamf)
varflag(VAR_beep,	"beep",		TRUE,		NULL)
varflag(VAR_bg_output,	"bg_output",	TRUE,		NULL)
varenum(VAR_binary_eol,	"binary_eol",	EOL_LF,		NULL,	enum_eol)
varflag(VAR_borg,	"borg",		TRUE,		NULL)
varenum(VAR_cecho,	"cecho",	0,		NULL,	enum_mecho)
varstr (VAR_cecho_attr,	"cecho_attr",	"Cgreen",	ch_attr)
varflag(VAR_cleardone,	"cleardone",	FALSE,		NULL)
varflag(VAR_clearfull,	"clearfull",	FALSE,		NULL)
varenum(VAR_async_conn,	"connect",	TRUE,		NULL,	enum_block)
varflag(VAR_defcompile,	"defcompile",	FALSE,		NULL)
varenum(VAR_emulation,	"emulation",	EMUL_ANSI_ATTR,	NULL,	enum_emul)
varstr (VAR_error_attr,	"error_attr",	NULL,		ch_attr)
varflag(VAR_expand_tabs,"expand_tabs",	TRUE,		NULL)
varflag(VAR_expnonvis,	"expnonvis",	FALSE,		ch_expnonvis)
varflag(VAR_gag,	"gag",		TRUE,		NULL)
varenum(VAR_async_name,	"gethostbyname",TRUE,		NULL,	enum_block)
varint (VAR_gpri,	"gpri",		0,		NULL)
varflag(VAR_hilite,	"hilite",	TRUE,		NULL)
varstr (VAR_hiliteattr,	"hiliteattr",	"B",		ch_attr)
varpos (VAR_histsize,	"histsize",	1000,		NULL)
varflag(VAR_hook,	"hook",		TRUE,		NULL)
varint (VAR_hpri,	"hpri",		0,		NULL)
varenum(VAR_iecho,	"iecho",	0,		NULL,	enum_mecho)
varstr (VAR_iecho_attr,	"iecho_attr",	"Cred",		ch_attr)
varstr (VAR_info_attr,	"info_attr",	NULL,		ch_attr)
varflag(VAR_insert,	"insert",	TRUE,		NULL)
varflag(VAR_interactive,"interactive",	-1,		NULL)
varpos (VAR_isize,	"isize",	3,		ch_visual)
varflag(VAR_istrip,	"istrip",	FALSE,		NULL)
varstr (VAR_kbnum,	"kbnum",	NULL,		NULL)
varflag(VAR_kecho,	"kecho",	FALSE,		NULL)
varstr (VAR_kecho_attr,	"kecho_attr",	NULL,		ch_attr)
varflag(VAR_keepalive,	"keepalive",	TRUE,		tog_keepalive)
varflag(VAR_keypad,	"keypad",	FALSE,		tog_keypad)
varstr (VAR_kprefix,	"kprefix",	NULL,		NULL)
varflag(VAR_login,	"login",	TRUE,		NULL)
varflag(VAR_lp,		"lp",		FALSE,		tog_lp)
varflag(VAR_lpquote,	"lpquote",	FALSE,		ch_lpquote)
vartime(VAR_maildelay,	"maildelay",	60,0,		ch_maildelay)
varenum(VAR_matching,	"matching",	1,		NULL,	enum_match)
varint (VAR_max_hook,	"max_hook",	1000,		NULL)
varint (VAR_max_instr,	"max_instr",	1000000,	NULL)
varint (VAR_max_kbnum,	"max_kbnum",	999,		NULL)
varint (VAR_max_recur,	"max_recur",	100,		NULL)
varint (VAR_max_trig,	"max_trig",	1000,		NULL)
#if HAVE_MCCP
varflag(VAR_mccp,	"mccp",		TRUE,		NULL)
#else
varenum(VAR_mccp,	"mccp",		FALSE,		NULL,	enum_off)
#endif
varenum(VAR_mecho,	"mecho",	0,		NULL,	enum_mecho)
varstr (VAR_mecho_attr,	"mecho_attr",	NULL,		ch_attr)
varenum(VAR_meta_esc,	"meta_esc",	META_NONPRINT,	NULL,	enum_meta)
varint (VAR_msdp_dbg,	"msdp_dbg",	0,		NULL)
varflag(VAR_more,	"more",		FALSE,		tog_more)
varstr (VAR_mprefix,	"mprefix",	"+",		NULL)
varflag(VAR_oldslash,	"oldslash",	TRUE,		NULL)
varflag(VAR_optimize,	"optimize",	TRUE,		NULL)
varflag(VAR_pedantic,	"pedantic",	FALSE,		NULL)
varstr (VAR_prompt_sec,	"prompt_sec",	NULL,		obsolete_prompt)
varstr (VAR_prompt_usec,"prompt_usec",	NULL,		obsolete_prompt)
vartime(VAR_prompt_wait,"prompt_wait",	0,250000,	NULL)
varstr (VAR_proxy_host,	"proxy_host",	NULL,		NULL)
varstr (VAR_proxy_port,	"proxy_port",	"23",		NULL)
vartime(VAR_ptime,	"ptime",	1,0,		NULL)
varflag(VAR_qecho,	"qecho",	FALSE,		NULL)
varstr (VAR_qecho_attr,	"qecho_attr",	NULL,		ch_attr)
varstr (VAR_qprefix,	"qprefix",	NULL,		NULL)
varflag(VAR_quiet,	"quiet",	FALSE,		NULL)
varflag(VAR_quitdone,	"quitdone",	FALSE,		NULL)
varflag(VAR_redef,	"redef",	TRUE,		NULL)
varint (VAR_refreshtime,"refreshtime",	100000,		NULL)
varflag(VAR_scroll,	"scroll",	FALSE,		ch_visual)
varflag(VAR_secho,	"secho",	FALSE,		NULL)
varstr (VAR_secho_attr,	"secho_attr",	NULL,		ch_attr)
varflag(VAR_shpause,	"shpause",	TRUE,		NULL)
varpos (VAR_sidescroll,	"sidescroll",	999999,		NULL)
varint (VAR_sigfigs,	"sigfigs",	15,		NULL)
varflag(VAR_snarf,	"snarf",	FALSE,		NULL)
varflag(VAR_sockmload,	"sockmload",	FALSE,		NULL)
varstr (VAR_sprefix,	"sprefix",	NULL,		NULL)
varstr (VAR_stat_attr,	"status_attr",	NULL,		ch_status_attr)
varstr (VAR_stat_fields,"status_fields",NULL,		ch_status_fields)
varpos (VAR_stat_height,"status_height",1,		ch_status_height)
varstr (VAR_stat_pad,	"status_pad",	"_",		update_status_line)
varstr (VAR_stint_clock,"status_int_clock",NULL,	ch_status_int)
varstr (VAR_stint_more,	"status_int_more",NULL,		ch_status_int)
varstr (VAR_stint_world,"status_int_world",NULL,	ch_status_int)
varenum(VAR_sub,	"sub",		SUB_KEYWORD,	NULL,	enum_sub)
varpos (VAR_tabsize,	"tabsize",	8,		NULL)
varflag(VAR_telopt,	"telopt",	FALSE,		NULL)
varenum(VAR_textdiv,	"textdiv",	TRUE,		NULL,	enum_textdiv)
varstr (VAR_textdiv_str,"textdiv_str",	"=====",	NULL)
varstr (VAR_tfhost,	"tfhost",	NULL,		NULL)
varstr (VAR_time_format,"time_format",	"%H:%M",	NULL)
varflag(VAR_virtscreen,	"virtscreen",	TRUE,		undocumented_var)
varflag(VAR_visual,	"visual",	-1,		ch_visual)
varflag(VAR_warn_5keys, "warn_5keys",	TRUE,		NULL)
varflag(VAR_warn_curly_re,"warn_curly_re",TRUE,		NULL)
varflag(VAR_warn_def_B,	"warn_def_B",	TRUE,		NULL)
varflag(VAR_warn_status,"warn_status",	TRUE,		NULL)
varstr (VAR_warning_attr,"warning_attr",NULL,		ch_attr)
varflag(VAR_watchdog,	"watchdog",	FALSE,		NULL)
varflag(VAR_watchname,	"watchname",	FALSE,		NULL)
varstr (VAR_wordpunct,	"wordpunct",	"_",		NULL)
varflag(VAR_wrap,	"wrap",		TRUE,		ch_wrap)
varflag(VAR_wraplog,	"wraplog",	FALSE,		NULL)
varint (VAR_wrappunct,	"wrappunct",	10,		ch_wrap)
varint (VAR_wrapsize,	"wrapsize",	0,		ch_wrap)
varint (VAR_wrapspace,	"wrapspace",	4,		ch_wrap)

